### PR TITLE
Add MapDatasetOnOff

### PR DESCRIPTION
**Pull request by luca-giunti**
Implement a map dataset containing on-off counts and acceptance.

Some tests: https://github.com/luca-giunti/share/blob/master/ONOFF_test.ipynb

### CODE REVIEW

**Comment by cdeil** [comment]
@luca-giunti  @adonath @registerrier - Any chance to reduce the amount of code (e.g. by using extending methods in the parent classes, or using helper functions called from both map dataset classes)?

Just a thought. I'm unsubscribing now.

File: gammapy/cube/fit.py [hunk 1]
```diff
@@ -900,6 +900,622 @@ def to_image(self, spectrum=None, keepdims=True):
...
+        self.model = model
+        self.name = name
+
+        if mask_safe is None:
+            mask_safe = np.ones(self.data_shape, dtype=bool)
+
+        self.mask_safe = mask_safe
+        self.gti = gti
+
+    def __str__(self):
```
[the remark this answers is not in the document]
Comment (luca-giunti) [reply gammapy/cube/fit.py]: OK. Notice that this means that also the `.npred()` method from the `MapDataset` is called, which is fine if we set `background_model=None` on the `MapDatasetOnOff`. I guess I can also eliminate the `.npred_sig()` on the `MapDatasetOnOff` and just use the `.npred()` inherited from the `MapDataset`...

File: gammapy/cube/fit.py [hunk 2]
```diff
@@ -900,6 +900,622 @@ def to_image(self, spectrum=None, keepdims=True):
...
+        if self.edisp and other.edisp:
+            if isinstance(self.edisp, EDispMap) and isinstance(other.edisp, EDispMap):
+                self.edisp.stack(other.edisp)
+            else:
+                raise ValueError("Stacking of edisp kernels not supported")
+
+        if self.gti and other.gti:
+            self.gti = self.gti.stack(other.gti).union()
+
+    def likelihood(self):
```
[the remark this answers is not in the document]
Comment (adonath) [reply gammapy/cube/fit.py]: Sorry, of course you are right. Just leave as is...

File: gammapy/cube/fit.py [hunk 3]
```diff
@@ -900,6 +900,622 @@ def to_image(self, spectrum=None, keepdims=True):
...
+        edisp_map = Map.from_geom(geom_migra, unit="")
+        loc = migra_axis.edges.searchsorted(1.0)
+        edisp_map.data[:, loc, :, :] = 1.0
+        edisp = EDispMap(edisp_map, exposure_irf)
+
+        geom_rad = geom_irf.to_image().to_cube([rad_axis, energy_axis])
+        psf_map = Map.from_geom(geom_rad, unit="sr-1")
+        psf = PSFMap(psf_map, exposure_irf)
+
+        return cls(
```
[the remark this answers is not in the document]
Comment (adonath) [reply gammapy/cube/fit.py]: The change in `MapDataset` is already part of #2479, so no need to adapt it here...

File: gammapy/cube/fit.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (adonath) [reply gammapy/cube/fit.py]: Yes, I think this would be preferred solution for now, just do avoid the code duplication.

**Comment by luca-giunti** [comment]
Thanks @adonath for the review!

I addressed most of your comments. However I get 2 errors which after spending sufficient time I don't understand, maybe you might have an idea... one occurs in the test for the `.to_hdulist()` method and the other in the test of the stacking.

The error I get in the stacking part is especially puzzling. The strange behavior I get is the following: https://github.com/luca-giunti/share/blob/master/Atest.ipynb

I guess I am not making a right use of `super()`...

**Comment by adonath** [comment]
Thanks @luca-giunti, I'll take a look now...

**Comment by adonath** [comment]
@luca-giunti The usage of `super()` was basically correct. In https://github.com/gammapy/gammapy/pull/2481/commits/e347057c75e0b1298b50d6526d8efbe46906f8de I adapted the I/O test. The special behaviour there was that the test date are actually images, so no `_BANDS` HDUs are written. Once I adapted the reference HDU names the tests passed. 

Taking a look at the implementation of the `.stack()` method I'm not sure it's correct (at least it seems different from what we've done in `SpectrumDatasetOnOff.stack()` ). @registerrier Could you maybe check the implementations for consistency?

File: gammapy/cube/fit.py [hunk 4]
```diff
@@ -901,6 +901,404 @@ def to_image(self, spectrum=None, keepdims=True):
...
+            Dataset to be stacked with this one.
+        """
+        super().stack(other)
+
+        if self.alpha and other.alpha and self.counts_off and other.counts_off:
+            self.acceptance_off.data = 1 / (
+                self.alpha.data * self.counts_off.data
+                + other.alpha.data * other.counts_off.data
+            )
+            self.acceptance.data = np.ones(self.data_shape)
```
Comment (registerrier) [inline gammapy/cube/fit.py]: This looks incorrect indeed. 
- mask should be applied to compute the average `alpha`
- the average alpha has to be divided by the total OFF

**Comment by luca-giunti** [comment]
Thanks @adonath and @registerrier !

Regarding the stacking, indeed there were mistakes. However, this issue that I encountered in stacking a dataset with an identical one still persists... and moreover, it is not related to the `.stack()` in the `MapDatasetOnOff` but with the one in the `MapDataset`. 

For some reason, the `other` dataset is changed as well as the `self`...

Here I reproduced the issue, using gammapy-0.14 (it is there also in the last dev...): https://github.com/luca-giunti/share/blob/master/MapDataset_stacking.ipynb

**Comment by adonath** [comment]
@luca-giunti Actually this behaviour is expected, when you stack a dataset with itself, because the data is modified in place. If `dataset.data` and `other.data` refer to the same numpy data buffer in memory, then once you change `dataset.data`, `other.data` is changed as well. I think in practice the is rarely a problem, but in you specific case, it appears because of the way you create the tests datasets.

So what you need to do is an explicit copy, of the first dataset before stacking e.g.like so:
```
dataset = MapDataset()

stacked = dataset.copy()
stacked.stack(dataset)
```

**Comment by luca-giunti** [comment]
Thanks @adonath. The stacking should be better now, and I also added a test for the stacking of cutouts

**Comment by adonath** [comment]
Thanks @luca-giunti!

**Pull request merged**